### PR TITLE
Add link to MAX_RETRY allocation explain message

### PR DESCRIPTION
**Pull request by matthewabbott**
Adds maximum number of retries exceeded [reference link](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-allocation-explain.html#_maximum_number_of_retries_exceeded) to the max_retry allocation explanations string. 

Adds more detail to documentation page describing that this was done to protect the cluster, but the real cause of the issue may now be gone and so allocation can be retried.

Also adds `POST` to the example `_cluster/reroute` API in the explanation because some customers would use `GET` and be confused why it didn’t work.

### CODE REVIEW

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 1]
```diff
@@ -195,17 +195,20 @@ primary shard that has reached the maximum number of allocation retry attempts.
         {
           "decider": "max_retry",
           "decision" : "NO",
-          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [/_cluster/reroute?retry_failed=true] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"
+          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed=true] to retry, and for more information, see https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-allocation-explain.html#_maximum_number_of_retries_exceeded, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"
```
Comment (DaveCTurner) [inline docs/reference/cluster/allocation-explain.asciidoc:198]: The API call in the message is actually `POST /_cluster/reroute?retry_failed&metric=none` - see `org.elasticsearch.cluster.routing.allocation.decider.MaxRetryAllocationDecider#RETRY_FAILED_API`.

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 2]
```diff
@@ -195,17 +195,20 @@ primary shard that has reached the maximum number of allocation retry attempts.
...
 // NOTCONSOLE
-
-If decider message indicates a transient allocation issue, use 
-<<cluster-reroute,the cluster reroute API>> to retry allocation. 
+This message indicates that the cluster was previously unable to
+allocate this shard and chose to put a hold on further attempts.
+This is done to avoid burdening the cluster with repeated requests that will fail.
+If no other `no` decisions are present, then the transient allocation issue
+that caused these failures has most likely been resolved, and you can use the
+<<cluster-reroute,the cluster reroute API>> to retry allocation.
```
Comment (DaveCTurner) [inline docs/reference/cluster/allocation-explain.asciidoc:211]: I think this will be confusing, there are normally always some `no` decisions e.g. for nodes in the wrong data tier.

Also I'd rather we used the imperative voice: "use the reroute API" rather than just suggesting "you can ...".

Finally there's a duplicate `the` (one inside the link and one outside).

File: server/src/main/resources/org/elasticsearch/common/reference-docs-links.json [hunk 3]
```diff
@@ -43,5 +43,6 @@
   "MAX_SHARDS_PER_NODE": "size-your-shards.html#troubleshooting-max-shards-open",
   "FLOOD_STAGE_WATERMARK": "fix-watermark-errors.html",
   "X_OPAQUE_ID": "api-conventions.html#x-opaque-id",
-  "FORMING_SINGLE_NODE_CLUSTERS": "modules-discovery-bootstrap-cluster.html#modules-discovery-bootstrap-cluster-joining"
+  "FORMING_SINGLE_NODE_CLUSTERS": "modules-discovery-bootstrap-cluster.html#modules-discovery-bootstrap-cluster-joining",
+  "ALLOCATION_EXPLAIN_MAX_RETRY": "cluster-allocation-explain.html#_maximum_number_of_retries_exceeded"
```
Comment (DaveCTurner) [inline server/src/main/resources/org/elasticsearch/common/reference-docs-links.json:47]: Please add a fixed `[[anchor-name]]` in the docs rather than using the `#_auto_generated` one which might change inadvertently.

**Review by DaveCTurner — COMMENTED** [review]
Some small comments. Also you need to run `./gradlew precommit` and fix up the issues.

File: server/src/main/resources/org/elasticsearch/common/reference-docs-links.json (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/resources/org/elasticsearch/common/reference-docs-links.json:47]: See https://github.com/elastic/elasticsearch/pull/113667 which forbids this.

Comment (stefnestor) [reply server/src/main/resources/org/elasticsearch/common/reference-docs-links.json:47]: [X-post example](https://github.com/elastic/elasticsearch/pull/113656#discussion_r1778933646).

File: docs/reference/cluster/allocation-explain.asciidoc (hunk 2 above)
Comment (stefnestor) [reply docs/reference/cluster/allocation-explain.asciidoc:211]: Brainstorming, I might say
> Elasticsearch queues shard allocation retries in batches. If there are long running or a high quantity of shard recoveries occurring within the cluster, this process may time out for some shards resulting in `MAX_RETRY`. This surfaces infrequently but is expected to prevent infinite retries which may impact cluster performance. When encountered, run <<cluster-reroute,the cluster reroute API>> to retry allocation.

File: docs/reference/cluster/allocation-explain.asciidoc (hunk 1 above)
Comment (stefnestor) [inline docs/reference/cluster/allocation-explain.asciidoc:198]: Until we solidify the URL I vote leaving it off. Then Github suggestion to apply Dave's comment I believe would appear as:

```suggestion
          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed&metric=none] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"
```

**Comment by matthewabbott** [comment]
Yikes about all those commits. I did not rebase this correctly.

File: docs/reference/cluster/allocation-explain.asciidoc (hunk 1 above)
Comment (matthewabbott) [reply docs/reference/cluster/allocation-explain.asciidoc]: Thanks, fixed. It’s funny, it looks like the API command as it appears in our documentation right now is actually different from the command that shows up in the in-cluster response message.

File: docs/reference/cluster/allocation-explain.asciidoc (hunk 2 above)
Comment (matthewabbott) [reply docs/reference/cluster/allocation-explain.asciidoc]: Changed this to 
> Elasticsearch queues shard allocation retries in batches. If there are long-running shard
recoveries or a high quantity of shard recoveries occurring within the cluster, this
process may time out for some shards, resulting in `max_retry`. This surfaces infrequently
but is expected to prevent infinite retries which may impact cluster performance. When
encountered, run the <<cluster-reroute,cluster reroute>> API to retry allocation.

Which is basically identical but I tweaked the wording on the second sentence because I thought it sounded a bit clearer that way, and also moved 'the' and 'API' out of the link per suggestion from @DaveCTurner 

Thanks!

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 4]
```diff
@@ -195,17 +196,19 @@ primary shard that has reached the maximum number of allocation retry attempts.
         {
           "decider": "max_retry",
           "decision" : "NO",
-          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [/_cluster/reroute?retry_failed=true] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"
+          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed&metric=none] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"
```
Comment (DaveCTurner) [inline docs/reference/cluster/allocation-explain.asciidoc:199]: `org.elasticsearch.cluster.routing.allocation.decider.MaxRetryAllocationDecider#RETRY_FAILED_API` is `POST /_cluster/reroute?retry_failed` in v9, the `metric` parameter is deprecated & ignored. But in v8.x this is correct.

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 5]
```diff
@@ -195,17 +196,19 @@ primary shard that has reached the maximum number of allocation retry attempts.
...
   ]
 }
 ----
 // NOTCONSOLE
-
-If decider message indicates a transient allocation issue, use
-the <<cluster-reroute,cluster reroute>> API to retry allocation.
+Elasticsearch queues shard allocation retries in batches. If there are long-running shard
+recoveries or a high quantity of shard recoveries occurring within the cluster, this
+process may time out for some shards, resulting in `max_retry`. This surfaces infrequently
```
Comment (DaveCTurner) [inline docs/reference/cluster/allocation-explain.asciidoc:209]: This isn't true, there's no timeout in play here. You need to get 5 genuine failures in a row before you see this.

File: x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/math/HypotEvaluator.java [hunk 6]
```diff
@@ -13,29 +13,31 @@
...
 import org.elasticsearch.xpack.esql.core.tree.Source;
-import org.elasticsearch.xpack.esql.expression.function.Warnings;
 
 /**
  * {@link EvalOperator.ExpressionEvaluator} implementation for {@link Hypot}.
  * This class is generated. Do not edit it.
  */
 public final class HypotEvaluator implements EvalOperator.ExpressionEvaluator {
-  private final Warnings warnings;
+  private final Source source;
```
Comment (DaveCTurner) [inline x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/math/HypotEvaluator.java:25]: These changes seem unrelated?

**Review by DaveCTurner — CHANGES_REQUESTED** [review]

File: x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/math/HypotEvaluator.java (hunk 6 above)
Comment (matthewabbott) [reply x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/math/HypotEvaluator.java]: It looks like these are the changes from https://github.com/elastic/elasticsearch/commit/30ff4741c674532c832c53a3bd31aa307dce8a95 (Commit `30ff474`), I think? I’m not sure why they’re showing up here, but presumably it’s because I did something wrong when merging upstream/main into my branch. Hopefully this most merge removes them from the listed changes.

File: docs/reference/cluster/allocation-explain.asciidoc (hunk 4 above)
Comment (matthewabbott) [reply docs/reference/cluster/allocation-explain.asciidoc]: Tweaked this accordingly to just be `POST /_cluster/reroute?retry_failed`

**Review by DaveCTurner — APPROVED** [review]
LGTM
